src/psyclone/psyir/transformations/intrinsics/dotproduct2code_trans.py

```diff
@@ -0,0 +1,321 @@
+                        f"The DotProduct2CodeTrans transformation only "
+                        f"supports the transformation of a dotproduct "
+                        f"intrinsic with an argument not an array "
+                        f"slice if the argument is for the 1st dimension "
+                        f"of the array and is for the full range of that "
+                        f"dimension, but found {self._writer(arg)} in "
+                        f"{self._writer(node)}.")
+
+        # Both arguments should be real (as other intrinsic datatypes
+        # are not suported).
+        for arg in node.children:
+            if arg.symbol.datatype.intrinsic != ScalarType.Intrinsic.REAL:
+                raise TransformationError(
+                    f"The DotProduct2CodeTrans transformation only supports "
+                    f"arrays of real data, but found {self._writer(arg)} of "
+                    f"type {arg.symbol.datatype.intrinsic.name} in "
+                    f"{self._writer(node)}.")
+
+        # Check whether _get_array_bound raises an exception
+        _get_array_bound(node.children[0], node.children[1], self._writer)
+
+    def apply(self, node, options=None):
+        '''Apply the DOT_PRODUCT intrinsic conversion transformation to the
+        specified node. This node must be a DOT_PRODUCT
+        BinaryOperation.  If the transformation is successful then an
+        assignment which includes a DOT_PRODUCT BinaryOperation node
+        is converted to equivalent inline code.
+
+        :param node: a DOT_PRODUCT Binary-Operation node.
+        :type node: :py:class:`psyclone.psyir.nodes.BinaryOperation`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
+
+        '''
+        self.validate(node)
+
+        assignment = node.ancestor(Assignment)
+        vector1 = node.children[0]
+        vector2 = node.children[1]
+        symbol_table = node.ancestor(Routine).symbol_table
+
+        # Create new i loop iterator.
+        i_loop_symbol = symbol_table.new_symbol(
+            "i", symbol_type=DataSymbol, datatype=INTEGER_TYPE)
+
+        # Create temporary result variable. There is an assumption
+        # here that the DOT_PRODUCT Operator returns a PSyIR real
+        # type. This might not be what is wanted (e.g. the args might
+        # PSyIR integers), or there may be errors (arguments are of
+        # different types) but this can't be checked as we don't have
+        # the appropriate methods to query nodes (see #658).
+        symbol_res_var = symbol_table.new_symbol(
+            "res_dot_product", symbol_type=DataSymbol, datatype=REAL_TYPE)
+
+        # Replace operation with the temporary result variable.
+        result_ref = Reference(symbol_res_var)
+        node.replace_with(result_ref)
+
+        # Create "vector1(i)"
+        vector1_dims = [Reference(i_loop_symbol)]
+        if len(vector2.children) > 1:
```

Comment:
Should be `vector1.children` - implies a missing test?

Reply:
Well spotted. I've added a test that would pick this up (`test_apply_extra_dims_sizes`)

Reply:
Hopefully this is done.